### ISSUE TITLE
scatter ignores marker cycler

### ISSUE THREAD

**Issue by ianhbell** [matplotlib/matplotlib#15105]
### Bug report

**Bug summary**

The marker cycler does not appear to work with scatter

**Code for reproduction**

```python
import matplotlib.pyplot as plt
fig, ax = plt.subplots(1, 1)
ax.set_prop_cycle('marker', ['x','o','d','s'])
x= [1, 2, 3]
for yoffset in 1,2,3,4:
    ax.scatter(x, [_y+yoffset for _y in x])
plt.show()
#
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/11966765/63551395-4b8a5f00-c4f2-11e9-8064-30f7d883a266.png)

**Expected outcome**

A cycling marker for each call to scatter

**Matplotlib version**
  * Operating system: windows
  * Matplotlib version: 3.1.1
  * Matplotlib backend : Qt5Agg
  * Python version: 3.7

conda-forge

**Comment by tacaswell** [comment]
To be clear, this is a feature request, I do not think this has ever worked this way.

**Comment by ianhbell** [comment]
No problem, of course there are workarounds, but it would be far nicer to not need to workaround.  Can you perhaps point me to where to look?  I could see if I could contribute to this.  Seems easy(ish)

**Comment by WeatherGod** [comment]
When I implemented support for property cyclers, I approached it as
extending any existing color cycling into the propery cycling concepts.
Because `scatter()` does not do color cycling, I never added property
cycling support to it. I am not exactly sure this should get property
cycling support, but if scatter() does, then it definitely should be
restricted to only certain properties.

**Comment by jklymak** [comment]
A reasonable workaround?  Or do you need `scatter` for some reason?

```python
fig, ax = plt.subplots(1, 1)
ax.set_prop_cycle('marker', ['x','o','d','s'])
x= [1, 2, 3]
for yoffset in 1,2,3,4:
    ax.plot(x, [_y+yoffset for _y in x], lw=0)
plt.show()
```

**Comment by ianhbell** [comment]
Yeah, I really do need a scatter because I also want to color the markers by a third variable, and use the marker to distinguish datasets.  I stripped down my example for filing the issue.

**Comment by jklymak** [comment]
I think you can still use a cycler manually: https://matplotlib.org/cycler/  

FWIW, I just always do this stuff manually, because I never want the cycle to loop anyhow.  i.e. 
```python
markers = ['d', 's', 'v', '^', 'o']`   
for n in range(ndatasets):
   scatter(data[n], c=colordata[n], markers[n])
```
No doubt that is what you meant by "various workarounds"

I'm skeptical that it will be "easy" to incorporate this into scatter without all sorts of confusion.  `scatter` is already a bit of a beast.

**Comment by ianhbell** [comment]
But I do want it to automatically cycle, which is the whole point of this feature request :)

**Comment by timhoffm** [comment]
-0.5 on using the cycler in `scatter`. As discussed above it could only be applied to some of the properties (e.g. color and size are already variable within the function). I also argue that it's far less common to plot multiple scatter variants than to plot multiple lines/markers (i.e. `plot()`).

If you really need it, you could write a wrapper like*:
~~~
def scatter_prop(ax, *args, **kwargs):
    if 'marker' not in kwargs:
        kwargs['marker'] = next(ax._get_lines.prop_cycler)['marker']
    ax.scatter(*args, **kwargs)
~~~

*admittedly has two disadvantages: 1. Changes the API from a method to a function. 2. must rely on the private `_get_lines` - we currently don't have a public API to access the cycler.

**Comment by ImportanceOfBeingErnest** [comment]
I don't understand the argument. `plot` cycles through all the parameters, like color, linewidth, markersize, marker; but currently `scatter` only cycles through colors. I don't think there is (or should be) any reason for this restriction, other than that it's not implemented.

**Comment by timhoffm** [comment]
Alright, I didn't realize we're already down that rabbit hole.

**Comment by WeatherGod** [comment]
scatter() does not cycle through anything. There is no restriction, color
cycling was never implemented because the color was color mappable. I don't
know any plotting library that color-cycles scatter().

I am not against adding cycling support for scatter(), but like I said
before, it would need to be restricted.

**Comment by ImportanceOfBeingErnest** [comment]
What I mean by cycling is this:

```
import numpy as np
import matplotlib.pyplot as plt

c = plt.cycler(color=["indigo", "crimson", "gold"],
               markersize=[10, 20, 30],
               marker=["o", "s", "^"])
plt.rcParams["axes.prop_cycle"] = c

fig, ax = plt.subplots(1, 1)

x= np.array([1, 2, 3])

for i in range(6):
    ax.scatter(x, x+i)

plt.show()
```

![image](https://user-images.githubusercontent.com/23121882/63658605-42c4b200-c7ac-11e9-8527-8e4834b978a3.png)

and if in that code you replace `ax.scatter(x, x+i)` by `ax.plot(x,x+i, linestyle="none")` you get

![image](https://user-images.githubusercontent.com/23121882/63658623-6be54280-c7ac-11e9-9cbe-904a1b32d4f1.png)

Other libraries may not cycle through colors, but instead *markers*, but at least it's useful that *something*  is changed from run to run to get a default distinguishability when plotting several scatters.

**Comment by WeatherGod** [comment]
I am completely surprised to see scatter() cycling colors. I was certain
that it didn't do that at all. Now I want to figure out when that was added.

**Comment by ImportanceOfBeingErnest** [comment]
In version 2.0. The [style changes](https://matplotlib.org/users/dflt_style_changes.html#scatter) says

> The following changes were made to the default behavior of scatter: [...]
If the color of the markers is not specified it will follow the property cycle, pulling from the 'patches' cycle on the Axes.

Looks like it came in via #6291

**Comment by timhoffm** [comment]
Obviously not. I'm still not convinced changing this behavior is worth the API break.

**Comment by timhoffm** [comment]
I revise my stance in the severeness of API breakage. Since markers are not included in the default property cycle, this only affects users that set marker cycling and use scatter. That’s probably rare as you currently would only set marker cycling if you have multiple plot() calls, so you would have multiple plot() calls and at least one scatter() .
